helix-core/src/main/java/org/apache/helix/controller/stages/TopStateHandoffReportStage.java

```diff
@@ -218,6 +217,8 @@ private void reportTopStateExistence(ResourceControllerDataProvider cache, Curre
       // we observed an entire top state handoff process
       reportSingleTopStateHandoff(cache, lastTopStateInstance, currentTopStateInstance,
```

Comment:
I feel like we should modify the if check to just use `if (lastTopStateInstance !=null)` and reuse the `reportSingleTopStateHandoff` function for both usecases instead by getting rid of the if statement that returns on `topStateInstance == currentTopStateInstance` in `reportSingleTopStateHandoff` function.

Reply:
+1 Why do need two branches here? Even if we simply remove this if condition as well as first check in reportSingleTopStateHandoff about same instance then also it will automatically take care of this case right?

Reply:
i can do that optimization but i had concern with this code: 
    // Make sure last top state instance has not bounced during cluster data cache refresh
    if (cache.getLiveInstances().containsKey(lastTopStateInstance)) {
      String lastTopStateSession =
          cache.getLiveInstances().get(lastTopStateInstance).getEphemeralOwner();
      // We need this null check as there are test cases creating incomplete current state
      if (cache.getCurrentState(lastTopStateInstance, lastTopStateSession).get(resourceName)
          != null) {
        startTime =
            cache.getCurrentState(lastTopStateInstance, lastTopStateSession).get(resourceName)
                .getStartTime(partition.getPartitionName());
        fromTopStateUserLatency =
            cache.getCurrentState(lastTopStateInstance, lastTopStateSession).get(resourceName)
                .getEndTime(partition.getPartitionName()) - startTime;
      }
    }

So i wanted to first make sure change works properly and then optimize it but code de-dup. i can combine the two into one change.

let me know what you would prefer. i prefer the current approach but i am ready to hear reviewers choice too.
thanks,

Reply:
Resolving as i updated the conditions in the other method.